### PR TITLE
Fix scrambling bugs #104

### PR DESCRIPTION
**Pull request by kraigher**
- Ensures ScrambleTeams() is not automatically called if IsScrambleEnabled is False
- Fix almost scamble message to take roundsToScramble configuration into account.
- Fix admin team scramble only done on T-win and not if more than 3 rounds in a row have been won.

### CODE REVIEW

File: RetakesPlugin/Modules/Managers/GameManager.cs [hunk 1]
```diff
@@ -66,40 +69,34 @@ public void AddScore(CCSPlayerController player, int score)
 
     private int _consecutiveRoundsWon;
 
-    public void TerroristRoundWin()
+    private void TerroristRoundWin()
     {
         _consecutiveRoundsWon++;
 
-        if (_consecutiveRoundsWon == _consecutiveRoundWinsToScramble)
+        bool shouldScrambleNow = _isScrambleEnabled && _consecutiveRoundsWon == _consecutiveRoundWinsToScramble;
```
Comment (B3none) [inline RetakesPlugin/Modules/Managers/GameManager.cs:76]: As the types of these variables can be inferred by the declaration, use var instead.

File: RetakesPlugin/RetakesPlugin.cs [hunk 2]
```diff
@@ -557,23 +557,7 @@ public HookResult OnRoundPreStart(EventRoundPrestart @event, GameEventInfo info)
...
-
-            case CsTeam.Terrorist:
-                Helpers.WriteLine($"{LogPrefix}Calling TerroristRoundWin()");
-                _gameManager.TerroristRoundWin();
-                Helpers.WriteLine($"{LogPrefix}TerroristRoundWin call complete");
-                break;
-        }
-
-        _gameManager.BalanceTeams();
+        _gameManager.OnRoundPreStart(_lastRoundWinner);
```
Comment (B3none) [inline RetakesPlugin/RetakesPlugin.cs:560]: Moving this logic into the gameManager is legit, nice! 👍

File: RetakesPlugin/Modules/Managers/GameManager.cs [hunk 3]
```diff
@@ -66,40 +69,34 @@ public void AddScore(CCSPlayerController player, int score)
...
-            {
-                Server.PrintToChatAll(
-                    $"{RetakesPlugin.MessagePrefix}{_translator["retakes.teams.almost_scramble", _consecutiveRoundsWon, _consecutiveRoundWinsToScramble - _consecutiveRoundsWon]}");
-            }
-            else
-            {
-                Server.PrintToChatAll(
-                    $"{RetakesPlugin.MessagePrefix}{_translator["retakes.teams.win_streak", _consecutiveRoundsWon]}");
-            }
+        else if (shouldAlmostScramble) {
```
Comment (B3none) [inline RetakesPlugin/Modules/Managers/GameManager.cs:88]: Please make sure to follow the existing project formatting

File: RetakesPlugin/Modules/Managers/GameManager.cs [hunk 4]
```diff
@@ -198,6 +195,28 @@ public void BalanceTeams()
         SetTeams(newTerrorists, newCounterTerrorists);
     }
 
+    public void OnRoundPreStart(CsTeam winningTeam) {
+        // Handle team swaps during round pre-start.
+        switch (winningTeam)
```
Comment (B3none) [inline RetakesPlugin/Modules/Managers/GameManager.cs:200]: Can we also keep the calls to the debug logs too

File: RetakesPlugin/Modules/Managers/GameManager.cs [hunk 5]
```diff
@@ -66,40 +69,34 @@ public void AddScore(CCSPlayerController player, int score)
...
-    public void TerroristRoundWin()
+    private void TerroristRoundWin()
     {
         _consecutiveRoundsWon++;
 
-        if (_consecutiveRoundsWon == _consecutiveRoundWinsToScramble)
+        bool shouldScrambleNow = _isScrambleEnabled && _consecutiveRoundsWon == _consecutiveRoundWinsToScramble;
+        int roundsLeftToScramble = _consecutiveRoundWinsToScramble - _consecutiveRoundsWon;
+        // Almost scramble if 1-2 rounds left to automatic scramble
+        bool shouldAlmostScramble = _isScrambleEnabled && 0 < roundsLeftToScramble && roundsLeftToScramble <= 2;
```
Comment (B3none) [inline RetakesPlugin/Modules/Managers/GameManager.cs:79]: I'm not a fan of having the constant value on the left of an expression, I think this is what you're after instead:
```c#
var shouldAlmostScramble = _isScrambleEnabled && roundsLeftToScramble > 0 && roundsLeftToScramble <= 2;
```

**Review by B3none — CHANGES_REQUESTED** [review]
Thanks for your first contribution 🚀

Overall it looks good, just some styling tweaks.

Are you also able to provide evidence that it was tested with a server full of players before I make the release?

If not I can merge it and ask people to test it for you, but it'll have to go into another branch. ~~I've created a branch called `scramble-updates` so you can just point it there and I'll ask one of the server owners running the plugin if they would like to test it out for us.~~ We send them the artifacts from the latest PR build.

File: RetakesPlugin/Modules/Managers/GameManager.cs (hunk 3 above)
Comment (kraigher) [reply RetakesPlugin/Modules/Managers/GameManager.cs:88]: Will fix, I forgot to run format on save

File: RetakesPlugin/Modules/Managers/GameManager.cs (hunk 1 above)
Comment (kraigher) [reply RetakesPlugin/Modules/Managers/GameManager.cs:76]: Sure, for bool I think it is kind of nice to see the type. But I will change it to fix your style.

File: RetakesPlugin/Modules/Managers/GameManager.cs (hunk 5 above)
Comment (kraigher) [reply RetakesPlugin/Modules/Managers/GameManager.cs:79]: Ok will fix

File: RetakesPlugin/Modules/Managers/GameManager.cs (hunk 4 above)
Comment (kraigher) [reply RetakesPlugin/Modules/Managers/GameManager.cs:200]: There where no calls to Helpers.WriteLine within the GameManager before? Do you want me to add granular logs inside GameManager or is it enough to have debug logs in RetakesPlugin.cs around the call to OnRoundPreStart?

By the way it seems the LogPrefix is added twice for the debug log as each call to Helpers.WriteLine adds the prefix by the caller but then the prefix is also added inside the method.

Comment (B3none) [reply RetakesPlugin/Modules/Managers/GameManager.cs:200]: For now we can just port the existing logs that were in the part of the `RetakesPlugin.cs` that was migrated over. The helper is statically called so it shouldn't be an issue. Most of the helpers that were added were to aid stability in the earlier development of the plugin. Don't worry about adding extras in unless it's not behaving as expected.

Comment (kraigher) [reply RetakesPlugin/Modules/Managers/GameManager.cs:202]: Ok I added the debug prints arround the call to GameManager.OnRoundPreStart now before I saw your reply? Should I move it inside arround the calls to CounterTerroristRoundWin etc?

Comment (kraigher) [reply RetakesPlugin/Modules/Managers/GameManager.cs:202]: By the way I will make a parallel PR on master to remove the double prefix.

**Comment by B3none** [comment]
You'll have to excuse me whilst I just work through the formatting workflow 👍

**Comment by B3none** [comment]
@kraigher merge the latest master, fix your conflicts and I'll see about finding someone to test it 👍

**Comment by kraigher** [comment]
I rebased on master now

**Comment by B3none** [comment]
I've bumped the version and asked people to test it to ensure it's good. Once we've got confirmation I'll merge the changes and make a release.

https://canary.discord.com/channels/1160907911501991946/1187531929831096391/1226126922816163902

**Comment by kraigher** [comment]
Great I will make a new PR after this is merged which implements the AutoTeamBalance flag described in #104

**Review by B3none — APPROVED** [review]

**Pull request merged**